apis/Google.Cloud.Datastore.V1/Google.Cloud.Datastore.V1/DatastoreDb.cs

```diff
@@ -83,6 +84,61 @@ public abstract class DatastoreDb
         public static DatastoreDb Create(string projectId, string namespaceId = "", DatastoreClient client = null) =>
             new DatastoreDbImpl(projectId, namespaceId, client ?? DatastoreClient.Create());
 
+        /// <summary>
+        /// Creates a <see cref="DatastoreDb"/> by either connecting to the production servers using the
+        /// default application credentials, or by connecting to a local emulator. The behavior depends on
+        /// the DATASTORE_EMULATOR_HOST and DATASTORE_PROJECT_ID environment variables, with further control
+        /// provided by the <paramref name="environmentBehavior"/> parameter.
+        /// </summary>
+        /// <param name="environmentBehavior">The way to behave in different environments. See the enum
+        /// definition for more details.</param>
+        /// <param name="projectId">The project ID to use when connecting to production servers. Ignored
+        /// when connecting to the emulator. Must not be null for production scenarios.</param>
+        /// <param name="namespaceId">The namespace ID to use in operations requiring a partition. Must not be null.</param>
+        /// <exception cref="InvalidOperationException">The environment was not configured appropriately
+        /// for the requested behavior.</exception>
+        /// <returns>A <see cref="DatastoreDb"/> operating on the specified partition, either connected
+        /// to production servers or the Datastore emulator.</returns>
+        public static DatastoreDb CreateWithEnvironment(EnvironmentBehavior environmentBehavior, string projectId, string namespaceId = "")
```

Comment:
I think I'd made `projectId` optional and then also check its value depending on the behaviour. I think this might make things more symetrical, since we would need `EmulatorOnlyThrowIfProduction` and `EmulatorOnlyIgnoreProduction`.

Reply:
We already only check its value based on whether we want to use it or not, but yes, we could make it optional.

EmulatorOnlyIgnoreProduction doesn't make sense to me as an option, because "production environment" means "no emulator environment variables set" which means we don't know how to talk to the emulator.

Reply:
I think "production environment" means "no emulator environment variables set" and "projectId set" (which we are checking _if_ we use it to create the `DatastoreDb`). So there's a chance that someone passes a value for projectId but ends up using the Emulator because the env variables are set, without it ever failing. I think I'd like it to fail in that case (unless EmulatorOnlyIgnoreProduction) or else I find it confussing specially because for using the production environment we do have the ignore and throw options. But this might just be me :)

Reply:
Hmm... it makes it harder if the production project ID is a constant though... it means there are then two things to change when we want to use the emulator, rather than only one. It may be worth chatting more about this over a VC. We certainly *could* fail if the `projectId` is specified but we only want to use the emulator.